File: apps/android/src/main/java/com/rustella/stellartrail/feature/gear/list/GearListViewModel.kt
package com.rustella.stellartrail.feature.gear.list

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.rustella.stellartrail.core.network.userMessage
import com.rustella.stellartrail.data.gear.GearRepositoryContract
import com.rustella.stellartrail.domain.gear.GearCategoriesResponse
import com.rustella.stellartrail.domain.gear.GearCategory
import com.rustella.stellartrail.domain.gear.GearSort
import com.rustella.stellartrail.domain.gear.GearStatsResponse
import com.rustella.stellartrail.domain.gear.GearStatus
import com.rustella.stellartrail.domain.gear.GearSummary
import com.rustella.stellartrail.domain.gear.GearTab
import com.rustella.stellartrail.domain.gear.GearTemplate
import com.rustella.stellartrail.domain.gear.ListGearsRequest
import com.rustella.stellartrail.feature.home.EMPTY_STATS
import kotlinx.coroutines.async
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch
import kotlinx.coroutines.supervisorScope

data class GearListUiState(
    val isLoggedIn: Boolean = false,
    val tab: GearTab = GearTab.AVAILABLE,
    val selectedCategory: GearCategory? = null,
    val selectedStatus: GearStatus? = null,
    val sort: GearSort = GearSort.CREATED_AT_DESC,
    val query: String = "",
    val categories: GearCategoriesResponse = GearCategoriesResponse(emptyList()),
    val stats: GearStatsResponse = EMPTY_STATS,
    val gears: List<GearSummary> = emptyList(),
    val templates: List<GearTemplate> = emptyList(),
    val nextCursor: String? = null,
    val loading: Boolean = false,
    val loadingMore: Boolean = false,
    val error: String? = null,
)

class GearListViewModel(private val repository: GearRepositoryContract) : ViewModel() {
    private val _state = MutableStateFlow(GearListUiState())
    val state: StateFlow<GearListUiState> = _state.asStateFlow()

    fun refresh(isLoggedIn: Boolean = true) {
        viewModelScope.launch {
            _state.update {
                it.copy(
                    isLoggedIn = isLoggedIn,
                    loading = true,
                    error = null,
                    gears = emptyList(),
                    templates = if (isLoggedIn) it.templates else emptyList(),
                    nextCursor = null,
                )
            }
            try {
                if (!isLoggedIn) {
                    val templates = repository.listTemplates().items
                    _state.update {
                        it.copy(
                            categories = GearCategoriesResponse(emptyList()),
                            stats = EMPTY_STATS,
                            gears = emptyList(),
                            templates = templates,
                        )
                    }
                    return@launch
                }
                val request = buildRequest(cursor = null)
                supervisorScope {
                    val categories = async { repository.listCategories(request.tab) }
                    val stats = async { repository.stats(request.tab) }
                    val list = async { repository.list(request) }
                    val templates = async { repository.listTemplates() }
                    val categoriesValue = categories.await()
                    val statsValue = stats.await()
                    val response = list.await()
                    val templateItems = templates.await().items
                    _state.update {
                        it.copy(
                            categories = categoriesValue,
                            stats = statsValue,
                            gears = response.items,
                            templates = templateItems,
                            nextCursor = response.nextCursor,
                        )
                    }
                }
            } catch (throwable: Throwable) {
                _state.update { it.copy(error = throwable.userMessage()) }
            } finally {
                _state.update { it.copy(loading = false) }
            }
        }
    }

    fun loadMore() {
        val cursor = _state.value.nextCursor ?: return
        if (_state.value.loadingMore || _state.value.loading || !_state.value.isLoggedIn) return
        viewModelScope.launch {
            _state.update { it.copy(loadingMore = true, error = null) }
            try {
                val response = repository.list(buildRequest(cursor))
                _state.update { it.copy(gears = it.gears + response.items, nextCursor = response.nextCursor) }
            } catch (throwable: Throwable) {
                _state.update { it.copy(error = throwable.userMessage()) }
            } finally {
                _state.update { it.copy(loadingMore = false) }
            }
        }
    }

    fun setTab(tab: GearTab) {
        _state.update { it.copy(tab = tab, selectedCategory = null, selectedStatus = null) }
        refresh(_state.value.isLoggedIn)
    }

    fun setCategory(category: GearCategory?) {
        _state.update { it.copy(selectedCategory = category) }
        refresh(_state.value.isLoggedIn)
    }

    fun setStatus(status: GearStatus?) {
        _state.update { it.copy(selectedStatus = status) }
        refresh(_state.value.isLoggedIn)
    }

    fun setSort(sort: GearSort) {
        _state.update { it.copy(sort = sort) }
        refresh(_state.value.isLoggedIn)
    }

    fun updateQuery(value: String) = _state.update { it.copy(query = value) }

    fun submitSearch() = refresh(_state.value.isLoggedIn)

    fun resetFilters() {
        _state.update {
            it.copy(
                selectedCategory = null,
                selectedStatus = null,
                sort = GearSort.CREATED_AT_DESC,
                query = "",
            )
        }
        refresh(_state.value.isLoggedIn)
    }

    fun archive(id: String) {
        if (!_state.value.isLoggedIn) return
        viewModelScope.launch {
            try {
                repository.archive(id)
                refresh(true)
            } catch (throwable: Throwable) {
                _state.update { it.copy(error = throwable.userMessage()) }
            }
        }
    }

    fun restore(id: String) {
        if (!_state.value.isLoggedIn) return
        viewModelScope.launch {
            try {
                repository.restore(id)
                refresh(true)
            } catch (throwable: Throwable) {
                _state.update { it.copy(error = throwable.userMessage()) }
            }
        }
    }

    fun delete(id: String) {
        if (!_state.value.isLoggedIn) return
        viewModelScope.launch {
            try {
                repository.delete(id)
                refresh(true)
            } catch (throwable: Throwable) {
                _state.update { it.copy(error = throwable.userMessage()) }
            }
        }
    }

    private fun buildRequest(cursor: String?): ListGearsRequest {
        val current = _state.value
        return ListGearsRequest(
            tab = current.tab,
            category = current.selectedCategory,
            status = current.selectedStatus,
            query = current.query.trim().takeIf { it.isNotEmpty() },
            sort = current.sort,
            limit = 20,
            cursor = cursor,
        )
    }
}

File: packages/apple/StellarTrailKit/Sources/StellarTrailKit/ViewModels/GearListViewModel.swift
import Foundation
import Combine

@MainActor
final class GearListViewModel: ObservableObject {
    struct State: Equatable {
        var loading = false
        var loadingMore = false
        var error: String?
        var isLoggedIn = false
        var tab: GearTab = .available
        var query = ""
        var selectedCategory: GearCategory?
        var selectedStatus: GearStatus?
        var sort: GearSort = .createdAtDesc
        var categories: [GearCategoryFilter] = []
        var stats = GearStatsResponse.empty
        var gears: [GearSummary] = []
        var templates: [GearTemplate] = []
        var atlasItems: [GearAtlasPublicItem] = []
        var atlasSubmissions: [GearAtlasSubmission] = []
        var nextCursor: String?
    }

    @Published private(set) var state = State()

    private let sessionStore: SessionStore
    private let gearRepository: any GearRepositorying
    private let gearAtlasRepository: any GearAtlasRepositorying
    private let contentRepository: any ContentRepositorying

    init(sessionStore: SessionStore, gearRepository: any GearRepositorying, gearAtlasRepository: any GearAtlasRepositorying, contentRepository: any ContentRepositorying) {
        self.sessionStore = sessionStore
        self.gearRepository = gearRepository
        self.gearAtlasRepository = gearAtlasRepository
        self.contentRepository = contentRepository
    }

    func load() async {
        state.loading = true
        state.error = nil
        state.isLoggedIn = sessionStore.isLoggedIn
        do {
            if sessionStore.isLoggedIn {
                let stats = try await gearRepository.stats(tab: state.tab)
                let categories = try await gearRepository.categories(tab: state.tab).items
                let response = try await gearRepository.list(currentRequest(cursor: nil))
                let submissions = try await gearAtlasRepository.mySubmissions(ListGearAtlasSubmissionsRequest(limit: 3)).items
                state.stats = stats
                state.categories = categories
                state.gears = response.items
                state.nextCursor = response.nextCursor
                state.templates = []
                state.atlasSubmissions = submissions
            } else {
                state.templates = try await contentRepository.gearTemplates().items
                state.atlasItems = try await gearAtlasRepository.list(ListGearAtlasRequest(limit: 3)).items
                state.gears = []
                state.categories = []
                state.stats = .empty
                state.nextCursor = nil
                state.atlasSubmissions = []
            }
            state.loading = false
            state.isLoggedIn = sessionStore.isLoggedIn
        } catch {
            state.loading = false
            state.error = error.localizedDescription
        }
    }

    func refreshWith(tab: GearTab? = nil, category: GearCategory? = nil, status: GearStatus? = nil, sort: GearSort? = nil, query: String? = nil) async {
        if let tab { state.tab = tab }
        if let category { state.selectedCategory = category }
        if let status { state.selectedStatus = status }
        if let sort { state.sort = sort }
        if let query { state.query = query }
        await load()
    }

    func clearCategory() async {
        state.selectedCategory = nil
        await load()
    }

    func clearStatus() async {
        state.selectedStatus = nil
        await load()
    }

    func loadMore() async {
        guard let cursor = state.nextCursor, !state.loadingMore else { return }
        state.loadingMore = true
        defer { state.loadingMore = false }
        do {
            let response = try await gearRepository.list(currentRequest(cursor: cursor))
            state.gears.append(contentsOf: response.items)
            state.nextCursor = response.nextCursor
        } catch {
            state.error = error.localizedDescription
        }
    }

    private func currentRequest(cursor: String?) -> ListGearsRequest {
        ListGearsRequest(tab: state.tab, category: state.selectedCategory, status: state.selectedStatus, q: state.query, sort: state.sort, limit: 20, cursor: cursor)
    }
}

@MainActor
final class GearDetailViewModel: ObservableObject {
    @Published private(set) var item: GearItem?
    @Published private(set) var submission: GearAtlasSubmission?
    @Published private(set) var submissions: [GearAtlasSubmission] = []
    @Published private(set) var loading = false
    @Published private(set) var submitting = false
    @Published private(set) var error: String?
    @Published private(set) var message: String?

    private let id: String
    private let repository: any GearRepositorying
    private let atlasRepository: any GearAtlasRepositorying

    init(id: String, repository: any GearRepositorying, atlasRepository: any GearAtlasRepositorying) {
        self.id = id
        self.repository = repository
        self.atlasRepository = atlasRepository
    }

    func load() async {
        loading = true
        error = nil
        message = nil
        do {
            async let gearItem = repository.get(id: id)
            async let atlasResponse = atlasRepository.mySubmissions(ListGearAtlasSubmissionsRequest(limit: 20))
            let loadedItem = try await gearItem
            let loadedSubmissionsResponse = try await atlasResponse
            let loadedSubmissions = loadedSubmissionsResponse.items
            item = loadedItem
            submissions = loadedSubmissions
            submission = loadedSubmissions.first { $0.sourceUserGearId == id || $0.name == loadedItem.name }
            loading = false
        } catch {
            loading = false
            self.error = error.localizedDescription
        }
    }

    func archive() async {
        do {
            try await repository.archive(id: id)
            await load()
        } catch {
            self.error = error.localizedDescription
        }
    }

    func restore() async {
        do {
            item = try await repository.restore(id: id)
        } catch {
            self.error = error.localizedDescription
        }
    }

    func submitToAtlas() async {
        guard !submitting else { return }
        submitting = true
        error = nil
        message = nil
        defer { submitting = false }
        do {
            let result = try await atlasRepository.submitGear(id: id)
            submission = result
            message = "已提交图鉴审核"
        } catch {
            self.error = error.localizedDescription
        }
    }
}
